scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala

```diff
@@ -132,58 +136,51 @@ class KRouter[A](
     network
       .ping(recToPing, Ping(uuidSource(), config.nodeRecord))
-      .map(_ => true)
+      .as(true)
       .onErrorHandle(_ => false)
 
   }
 
   def add(nodeRecord: NodeRecord[A]): Task[Unit] = {
-    info(s"Handling potential addition of candidate (${nodeRecord.id.toHex}, $nodeRecord)")
-    if (myself(nodeRecord.id)) {
-      Task.now(())
-    } else {
-      for {
-        toPing <- routerState.modify { current =>
-          val (_, bucket) = current.kBuckets.getBucket(nodeRecord.id)
-          if (bucket.size < config.k) {
-            (current.addNodeRecord(nodeRecord), None)
-          } else {
-            // the bucket is full, not update it but ping least recently seen node (i.e. the one at the head) to see what to do
-            val nodeToPing = current.nodeRecords(bucket.head)
-            (current, Some(nodeToPing))
+    Task.delay(logger.info(s"Handling potential addition of candidate (${nodeRecord.id.toHex}, $nodeRecord)")) *> {
+      if (myself(nodeRecord.id)) {
+        Task.now(())
+      } else {
+        for {
+          toPing <- routerState.modify { current =>
+            val (_, bucket) = current.kBuckets.getBucket(nodeRecord.id)
+            if (bucket.size < config.k) {
+              (current.addNodeRecord(nodeRecord), None)
+            } else {
+              // the bucket is full, not update it but ping least recently seen node (i.e. the one at the head) to see what to do
+              val nodeToPing = current.nodeRecords(bucket.head)
+              (current, Some(nodeToPing))
+            }
           }
-        }
-        result <- pingAndUpdateState(toPing, nodeRecord)
-      } yield result
+          result <- pingAndUpdateState(toPing, nodeRecord)
+        } yield result
+      }
     }
   }
 
   private def pingAndUpdateState(recordToPing: Option[NodeRecord[A]], nodeRecord: NodeRecord[A]): Task[Unit] = {
     recordToPing match {
       case None => Task.now(())
       case Some(nodeToPing) =>
-        for {
-          pingResult <- ping(nodeToPing)
-          _ <- routerState.update { current =>
-            if (pingResult) {
-              // if it does respond, it is moved to the tail and the other node record discarded.
-              current.touchNodeRecord(nodeToPing)
-            } else {
-              // if that node fails to respond, it is evicted from the bucket and the other node inserted (at the tail)
-              current.replaceNodeRecord(nodeToPing, nodeRecord)
-            }
-          }
-
-          _ <- Task.eval {
-            if (pingResult) {
-              info(
-                s"Moving ${nodeToPing.id} to head of bucket. Discarding (${nodeRecord.id.toHex}, $nodeRecord) as routing table candidate."
-              )
-            } else {
-              info(s"Replacing ${nodeToPing.id.toHex} with new entry (${nodeRecord.id.toHex}, $nodeRecord).")
-            }
-          }
-        } yield ()
+        ping(nodeToPing).ifM(
```

Comment:
This change has different semantics from the old code.
Old version was an: `atomic touch or replace based on ping result`, new version is: `atomic touch or atomic replace based on ping result`.  
Tbh i do not remember why it was important to keep it that way, but i would rather avoid semantic changing refactors, and if that part was overly synchronised just create task to improve that.

Reply:
There cannot be any difference between those, the ping result doesn't change based on whether you're checking inside or outside. What I wanted to get rid of is having to check it twice: once for the change, then once for the logging. That way they can easily get out of sync, if you have to repeat the same logic twice. The tests caught a problem when I used it wrong. And I'd rather not log inside `modify`, that just feels wrong.

Reply:
sure I would also not log in `modify`. So lets leave it as it is, I cannot think of reason why this approach would break

Reply:
Maybe `modify` uses `compareAndSet` under the hood and when logs were together with the mutation inside people saw them repeatedly, when there was a contention, that's why it was moved out.